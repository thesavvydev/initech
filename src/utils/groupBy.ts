export default function groupBy<T>(
  items: T[],
  keySelector: (item: T) => string
) {
  return items.reduce<{ [k: string]: T[] }>((dictionary, item) => {
    const itemKey = keySelector(item);
    dictionary[itemKey] = (dictionary[itemKey] ?? []).concat(item);
    return dictionary;
  }, {});
}
