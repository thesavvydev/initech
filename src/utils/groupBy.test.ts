import { describe, expect, test } from "vitest";
import groupBy from "./groupBy";

const array = [
  {
    id: 1,
    group: "a",
  },
  {
    id: 2,
    group: "b",
  },
  {
    id: 3,
    group: "a",
  },
  {
    id: 4,
    group: "a",
  },
  {
    id: 5,
    group: "a",
  },
  {
    id: 6,
    group: "b",
  },
];

describe("groupBy", () => {
  test("matches snapshot", () => {
    const result = groupBy(array, (arr) => arr.group);
    expect(result).toMatchSnapshot();
  });

  test("groups array by key selector", () => {
    const result = groupBy(array, (arr) => arr.group);
    expect(result).toMatchObject({
      a: [
        {
          id: 1,
          group: "a",
        },
        {
          id: 3,
          group: "a",
        },
        {
          id: 4,
          group: "a",
        },
        {
          id: 5,
          group: "a",
        },
      ],
      b: [
        {
          id: 2,
          group: "b",
        },
        {
          id: 6,
          group: "b",
        },
      ],
    });
  });
});
