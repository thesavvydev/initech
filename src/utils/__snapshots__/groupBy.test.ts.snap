// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`groupBy > matches snapshot 1`] = `
{
  "a": [
    {
      "group": "a",
      "id": 1,
    },
    {
      "group": "a",
      "id": 3,
    },
    {
      "group": "a",
      "id": 4,
    },
    {
      "group": "a",
      "id": 5,
    },
  ],
  "b": [
    {
      "group": "b",
      "id": 2,
    },
    {
      "group": "b",
      "id": 6,
    },
  ],
}
`;
